### PR TITLE
new flexiroute

### PR DESCRIPTION
**Pull request by drewhk**

### CODE REVIEW

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphFlexiRouteSpec.scala [hunk 1]
```diff
@@ -18,23 +18,14 @@ object GraphFlexiRouteSpec {
...
-        SameState
-      }
 
+    override def createRouteLogic(p: PortT): RouteLogic[T] = new RouteLogic[T] {
       // initally, wait for demand from all
-      override def initialState = State[T](DemandFromAll(out1, out2)) { (ctx, preferredOutput, element) ⇒
+      override def initialState = State[T](DemandFromAll(p)) { (ctx, preferredOutput, element) ⇒
         ctx.emit(preferredOutput, element)
-        emitToAnyWithDemand
+        SameState
```
Comment (rkuhn) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphFlexiRouteSpec.scala]: why this change in semantics?

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/FanInPorts.scala.template [hunk 2]
```diff
@@ -13,7 +13,7 @@ abstract class FanInPorts[O] extends Ports {
   final override def inlets: immutable.Seq[InPort[_]] = _inlets
   
   private var _inlets: List[InPort[_]] = Nil
-  protected def port[T](name: String) = {
+  protected def port[T](name: String): InPort[T] = {
```
Comment (rkuhn) [inline akka-stream/src/main/boilerplate/akka/stream/scaladsl/FanInPorts.scala.template]: thanks!

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphFlexiRouteSpec.scala (hunk 1 above)
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphFlexiRouteSpec.scala]: ah, didn't notice that this is all and the other is any!

Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphFlexiRouteSpec.scala]: fixed, and it still works. phew

File: akka-stream/src/main/scala/akka/stream/scaladsl/FlexiRoute.scala [hunk 3]
```diff
@@ -44,10 +28,11 @@ object FlexiRoute {
    * has been completed. `IllegalArgumentException` is thrown if
    * that is not obeyed.
    */
-  final case class DemandFrom(output: OutputHandle) extends DemandCondition
+  final case class DemandFrom(output: OutPort[_]) extends DemandCondition
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/scaladsl/FlexiRoute.scala]: So, I didn't add a type parameter to DemandCondition since the lowest common type for the bunched output ports (DemandFromAny/All) would be Nothing so I cannot do the same trick as you did in FlexiMerge (giving Any in those cases).

File: akka-stream/src/main/scala/akka/stream/scaladsl/FlexiRoute.scala [hunk 4]
```diff
@@ -138,8 +123,8 @@ object FlexiRoute {
      * The `onInput` function is called when an `element` was read from upstream.
      * The function returns next behavior or [[#SameState]] to keep current behavior.
      */
-    sealed case class State[Out](val condition: DemandCondition)(
-      val onInput: (RouteLogicContext[Out], OutputHandle, In) ⇒ State[_])
+    sealed case class State[Out](condition: DemandCondition)(
+      val onInput: (RouteLogicContext[Out], OutP, In) ⇒ State[_])
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/scaladsl/FlexiRoute.scala]: Here the types are not properly matched up: the `Out` is not fixed to the DemandCondition hence it does not enable better type-safety. If possible do it as in the FlexiMerge case, using proper type parameters on the DemandCondition. If that does not work then there is no point in using `OutPort[_]` at all and it should be changed to `OutP` everywhere.

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/scaladsl/FlexiRoute.scala]: I explained in another comment that the problem is that the lowest common type for a bunch of ports is likely Nothing :( So yes, maybe we can remove OutPort[_]
